Panache reactive based on Hibernate Reactive

FroMage opened the pull request:
Probably a few things left to fix, but at least all the tests pass, so a good candidate for running tests in CI.

FroMage committed:
Panache Reactive

FroMage committed:
HR: conditionally add the producers

To make sure we don't add producers that depend on something that ORM didn't add

FroMage committed:
junit5-mockito: support mocking Mutiny method default values

FroMage committed:
junit5-mockito: Do not mock the proxy classes but rather their superclass

FroMage commented:
Missing docs.

gavinking commented:
oh, nice, do you have a link to an example?

FroMage commented:
> oh, nice, do you have a link to an example?

https://github.com/quarkusio/quarkus/blob/592c9114898320c553436e898414e4a594fa68e8/integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/TestEndpoint.java#L56

But there's so much stuff in these tests that it really doesn't do justice to the API.

gavinking commented:
Right, I was more looking for an example app than tests.

FroMage commented:
I'll add a quickstart too

gsmet commented on extensions/panache/hibernate-reactive-panache/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,12 @@
+  - "panache"
+  - "hibernate"
+  - "jpa"
+  guide: "https://quarkus.io/guides/hibernate-orm-panache"
```
You won't have a specific guide?

FroMage replied:
We will, but there's no guide yet, as is the case for HR.

gsmet replied:
In that case, better remove the guide than pointing to an inappropriate one.

FroMage replied:
OK.

gsmet commented:
I suppose we want that for 1.7. What's the plan for the reviews?

FroMage commented:
Well, CI finally seems to pass, so let's get this out of draft.

gsmet commented:
What I was asking is who will review it?

Personally, I don't know much about Panache and I know nothing about reactive so I won't. Can you find some people who will get that reviewed and merged before CR1?

Sanne commented:
I can have a look, but "104 files changes" .. gosh I only have an hour I can spare today :)

Sanne commented:
@DavideD could you have a look as well? If you could focus on checking if the Hibernate Reactive APIs seem used correctly, I'll focus on integration risks.

Sanne commented on extensions/panache/panache-mock/src/main/java/io/quarkus/panache/mock/PanacheMock.java:
```diff
@@ -40,6 +41,13 @@ public static void mock(Class<?>... classes) {
 
+    public static void mock(MockSettings settings, Class<?>... classes) {
+        for (Class<?> klass : classes) {
+            mocks.computeIfAbsent(klass, v -> Mockito.mock(klass, settings));
```
What if the `settings` don't match from a previously cached mock you might have created for the same klass?

FroMage replied:
I guess it will override the mock, since they're keyed on class and not settings. That's fine.

Sanne replied:
and settings will not change?

FroMage replied:
Ah, you're right, OK.

FroMage replied:
Pfff, I'll just remove this method.

Sanne commented on extensions/panache/panache-mock/src/main/java/io/quarkus/panache/mock/PanacheMock.java:
```diff
@@ -40,6 +41,13 @@ public static void mock(Class<?>... classes) {
+        for (Class<?> klass : classes) {
+            mocks.computeIfAbsent(klass, v -> Mockito.mock(klass, settings));
+        }
+        IsMockEnabled = !mocks.isEmpty();
```
you have a race condition here; maybe it's not an issues as I don't expect many will invoke this method concurrecntly, but perhaps you'd still prefer to use a safer locking pattern?

I don't think this needs to be "high throughput" or high performance, so just synchronize it all?

FroMage replied:
OK

Sanne commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -0,0 +1,506 @@
+    }
+
+    public Uni<Void> persist(Mutiny.Session em, Object entity) {
+        if (!em.contains(entity)) {
```
What is the purpose of this check? And why should it not be persisted otherwise?

Sanne replied:
Also, why is this specific variation of the `persist` method having this check, but the other methods won't have the same semantics?

FroMage replied:
The purpose is to not call `persist` if it's already persisted. The ORM version has the same check. What happens if I call `persist` on an already persisted entity, otherwise?

FroMage replied:
The other variations of `persist` ultimately call this, so it's fine.

Sanne replied:
>  What happens if I call persist on an already persisted entity, otherwise?

Nothing, but at least we have a chance to verify the transaction is alive, extend timeouts, and things like that.

FroMage replied:
Well, the javadoc wasn't entirely clear about the semantics here.

Sanne commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -0,0 +1,506 @@
+    //
+    // Private stuff
+
+    public Uni<Mutiny.Session> getEntityManager() {
```
why name this "getEntityManager" when it doesn't return one? :)

Also: why returning an Uni<Session>, when you could just get the `Session` right away? opening a Session is non-blocking.

Sanne replied:
I mean, ask ArC for a `Mutiny.Session` rather than a `Uni<Mutiny.Session>`

FroMage replied:
I don't think there's any producer for `Mutiny.Session` since ArC doesn't do async injection, so not sure what you're talking about.

OK, apparently this was added after I did my PR. Fine, let's switch to that :)

Sanne replied:
BTW, opening a Session isn't a blocking operation.

Sanne commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -0,0 +1,506 @@
+        return entityManager;
+    }
+
+    public TransactionManager getTransactionManager() {
```
Should we remove this for now?

FroMage replied:
Yes

Sanne commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -0,0 +1,382 @@
+            return countQuery;
+        }
+
+        // try to generate a good count query from the existing query
```
All this stuff with string manipulation is extremely dodgy .. :)

Sanne replied:
I suppose you have much of the same in the non-reactive version of Panache? Could we avoid some copy paste for all this code?

FroMage replied:
Yes, good idea.

Sanne commented on extensions/panache/hibernate-reactive-panache/deployment/pom.xml:
```diff
@@ -0,0 +1,243 @@
+        </plugins>
+    </build>
+
+    <profiles>
```
Don't we normally have integration tests in the `/integration-tests`directory, in the root of the Quarkus source tree?

FroMage replied:
Those are for the hot-reload. They mirror what we have for ORM with Panache, except that one could run without docker because H2, which we don't support ATM for reactive. I could move them to the existing IT, but that'd be less symmetrical and I still hope we will support H2 for reactive at some point and get rid of this docker requirement.

Sanne commented on extensions/panache/hibernate-reactive-panache/deployment/src/main/java/io/quarkus/hibernate/reactive/panache/deployment/PanacheHibernateResourceProcessor.java:
```diff
@@ -0,0 +1,196 @@
+                boolean m1 = t.hasType(DOTNAME_UNI);
+                if (m1) {
+                    boolean m2 = t.getTypes().contains(SESSION_TYPE);
+                    System.err.println("REMOVE " + t + " m1: " + m1 + " m2: " + m2);
```
Why using `System.err` here?

FroMage replied:
Ooops.

Sanne commented on extensions/panache/hibernate-reactive-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/panache/test/DuplicateIdEntityTest.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Test
+    void shouldThrow() {
+        fail("A BuildException should have been thrown due to duplicate entity ID");
```
So Panache can't deal with composite IDs ?

FroMage replied:
IIRC this detects the case where you have multiple `@Id` annotations but no matching `@IdClass` which is always a blunder.

Sanne commented on extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/Panache.java:
```diff
@@ -0,0 +1,85 @@
+public class Panache {
+
+    /**
+     * Returns the current {@link EntityManager}
```
It's not an `EntityManager` .. in this case since it's public API I'd definitely fix it.

FroMage replied:
Yes, I've fixed it.

loicmathieu commented:
@FroMage it's a huge PR, I'll have a look begining of next week but I already have one question. 

During the MongoDB with Panache reactive implementation we decided to ship the reactive variant inside the same extension as the imperative one. So why not doing the same for Hibernate ?
And there is an official guideline for it by the way: https://github.com/quarkusio/quarkus/pull/5567#issuecomment-566010760 ;)

DavideD commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -0,0 +1,506 @@
+                        return Uni.createFrom().item(false);
+                    }
+                    return getEntityManager().flatMap(em -> em.remove(entity).map(v -> true));
+                });
```
You could use ifNull and IfNotNull here:
```
findById(entityClass, id)
    .onItem().ifNotNull().transformToUni( entity -> s.remove( entity ).onItem().transform( ignore -> Boolean.TRUE ))
    .onItem().ifNull().continueWith( Boolean.FALSE );
```

FroMage replied:
I can, but it's not clear to me that it'll be more readable :)

DavideD commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -0,0 +1,506 @@
+
+    public Uni<Void> persist(Mutiny.Session em, Object entity) {
+        if (!em.contains(entity)) {
+            return em.persist(entity).map(v -> null);
```
What about this:
```
		return em.persist( entity ).onItem().ignore().andContinueWithNull();
```

FroMage replied:
Yeah, I've found the `map(v -> null)` to be a lot shorter and to the point. I'd rather keep this :)

loicmathieu replied:
I first use `map(v -> null)` than Clement said that `.onItem().ignore().andContinueWithNull()` is better, I don't remember why (maybe it provides the item to be emitted ...)

DavideD commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -0,0 +1,506 @@
+
+    public Uni<Void> persist(Stream<?> entities) {
+        return getEntityManager().flatMap(
+                em -> Uni.combine().all().unis(entities.map(entity -> persist(em, entity)).collect(Collectors.toList()))
```
Similar for `delete` and `flush`

FroMage replied:
Oh I didn't know about this, I've used it, thanks. I really didn't expect `onItem` to lead me to an operation that worked on the completion of ALL items.

I do not see what you mean about `delete` and `flush` though, which don't use streams?

FroMage replied:
Except it doesn't work:

```java
// this works:
        List<Uni<Void>> uniList = entities.map(entity -> persist(session, entity)).collect(Collectors.toList());
        return Uni.combine().all().unis(uniList).discardItems();

// this does nothing:
        return Multi.createFrom().items(entities)
                     .map(entity -> persist(session, entity))
                     .onItem().ignoreAsUni();
```

@cescoffier is this a bug?

cescoffier replied:
`persist` returns a `Uni` right? To you need to `flatMap` instead of `map`.

FroMage replied:
Aha, yeah, makes a lot of sense.

DavideD reviewed (commented):
I left some comments about the use of the Mutiny API. They are not critical, though.
Looks good to me overall

loicmathieu reviewed (changes requested):
This overal looks good to me, I added some comments / questions.

There is a few points that raised some concerns:

- Are transactions supported ? Did they use `@Transactional` or `Panache.transaction()`? I didn't find any Hibernate reactive guide so I don't understand how transactions are supposed to works with it.
- hibernate-orm and hibernate-reactive are exclusive extensions right ? 
You cannot use both at the same time that why there is also two extensions for Panache ? 
But there is now a lot of code duplicated between the two (named query, select statement creation, ...). I know that there is already a lot of common modules inside Panache but maybe create a new one to hold the code duplicated between the orm and the reactive Panache extensions would be a good idea.
- In the reactive MongoDB with panache extension, all components are named prefixed by `Reactive` (ReactivePanacheEntity, ReactivePanacheQuery, ReactiveMongoOperation), as for Hibernate there is two extensions there is no issue but maybe it's better to do the same so the user clearly saw that he uses the reactive variant.

gavinking commented:
Right not, transactions in HR are always "resource local", that is, there's a HR API you use to control the Vert.x client database transaction.

gavinking commented:
> - In the reactive MongoDB with panache extension, all components are named prefixed by `Reactive` (ReactivePanacheEntity, ReactivePanacheQuery, ReactiveMongoOperation), as for Hibernate there is two extensions there is no issue but maybe it's better to do the same so the user clearly saw that he uses the reactive variant.

IMO this is a really terrible idea, and something we went out of our way to avoid in HR. Most users are using either one programming paradigm or the other, and aren't mixing together the two approaches in the same code, so there's never (or at least almost never) any major ambiguity about what sort of `Session` I have here.

It would be entirely obnoxious and repetitive to qualify every single API anyone uses in a reactive program with `Reactive`.

DavideD commented:
> It would be entirely obnoxious and repetitive to qualify every single API anyone uses in a reactive program with Reactive.

+1 

The main reason we use the `Reactive` prefix in Hibernate Reactive is to differentiate with the already existing classes or methods in ORM.

loicmathieu commented:
@gavinking 

> Most users are using either one programming paradigm or the other, and aren't mixing together the two approaches in the same code, so there's never (or at least almost never) any major ambiguity about what sort of Session I have here.

This was a choice we made some times ago for MongoDB, but the Hibernate story may be a little different as there is two extensions as opposite to MongoDB where the same extension provides both imperative and reactive client. 
So maybe having two extensions for Hibernate with Panache and only one for MongoDB with Panache is OK. 
Or maybe we need to split the MongoDB extension in two as I must agree that usually people are using imperative OR reactive and mixing both is not a good idea.

gavinking commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -0,0 +1,382 @@
+            jpaQuery.setLockMode(LockModeConverter.convertToLockMode(lockModeType));
+        }
+
+        if (hints != null) {
```
Alternatively, we should consider simply not supporting hints at all in Panache. They're really there only to allow use of vendor-specific features via the JPA-standard APIs, but they're untypesafe and inconvenient.

I would argue that it would be better if Panache offered typesafe APIs for Hibernate-specific features. As far as I understand, support for other JPA implementations is a non-goal.

FroMage replied:
They will be, let's throw.

gavinking replied:
> They will be, let's throw.

Seriously? you're going to make, say, Eclipselink work in Quarkus?

FroMage replied:
No, I meant hints will be supported, because I thought they would (and github didn't show me your answer at the time, so I was replying to Loïc).

I don't mind if we replace hint support if you tell me HR will not support them.

Sanne replied:
I'd love to fully eradicate support for hints from ORM, and have a proper API instead.

Of course we can't because of specs, but we could introduce an internal configurtion setting to at least ignore them for Hibernate Reactive (and maybe even for ORM classic when within Quarkus).
Beyond API benefits of more explicit options, dealing with them is also highl inefficient.

gavinking replied:
> I don't mind if we replace hint support if you tell me HR will not support them.

Yeah but I'm advocating removing them from the regular (non-reactive) API as well. Or rather, representing the specific hints that Hibernate supports in a type safe way, just like what I've done in certain specific cases in the HR API.

gavinking commented:
What I'm saying is I think you should consider revisiting that decision. It implicitly makes reactive a second-class citizen, and is a tax on anyone who actually wants to use the reactive APIs.

gavinking commented:
> The main reason we use the `Reactive` prefix in Hibernate Reactive is to differentiate with the already existing classes or methods in ORM.

But to be clear we only do it in internal implementation classes and interfaces, and some extension points. We don't do it for the APIs that the program interacts with all the time.

Sanne commented:
> * hibernate-orm and hibernate-reactive are exclusive extensions right ?
>   You cannot use both at the same time that why there is also two extensions for Panache ?

To be clear, you can't use them both _today_ but that limitation is meant to be temporary. It's just a consequence of the current limitation that Quarkus can boot a single persistence unit.

So that's not the reason to have two extensions; the real reason is that they have different dependencies - and as others said, people should be able to choose one (or both, eventually) without necessarily having to pull in a load of dependencies which aren't relevant to them.

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -0,0 +1,382 @@
+                    AbstractJpaOperations.bindParameters(countQuery, (Map<String, Object>) paramsArrayOrMap);
+                else
+                    AbstractJpaOperations.bindParameters(countQuery, (Object[]) paramsArrayOrMap);
+                // FIXME: question about caching the result here
```
Well, I also cache the Uni, but you don't cache its result. It could be cached or recreated every time.

loicmathieu replied:
Hum, good point.
There is a cache method on Uni it should be use to avoid this.

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -0,0 +1,382 @@
+    }
+
+    private void checkPagination() {
+        // FIXME: turn into Uni
```
Well, it also means we throw in situations where it'd be better to return a failed uni.

loicmathieu replied:
Hum, in fact I did it this way because ... you did it this way in your first implementation of Panache Rx so do as you want I'll update the MongoDB part later if needed ;)
I agree that you don't expect reactive code to throw, but this particular case is a developement error (like a NPE) so I think you can throw

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -0,0 +1,382 @@
+        page(page.first());
+    }
+
+    // FIXME: what do we do for this one?
```
OK.

FroMage commented on extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/PanacheQuery.java:
```diff
@@ -0,0 +1,264 @@
+ * @author Stéphane Épardaud
+ * @param <Entity> The entity type being queried
+ */
+public interface PanacheQuery<Entity> {
```
I really feel like `Optional` doesn't belong where we deal with `Uni` which have their own way to return empty results. I don't think we should add them.

loicmathieu replied:
Maybe ask some light from the munity expert, I'm OK if we don't provide them (I'll deprecate them in MongoDB then) as soon as we explain somewhere in the doc that we don't provide them because of Mutiny or someone may ask to add them

FroMage commented on extensions/panache/hibernate-reactive-panache/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,12 @@
+  guide: "https://quarkus.io/guides/hibernate-orm-panache"
+  categories:
+  - "data"
+  status: "stable"
```
Yeah, absolutely, thanks.

FroMage commented on integration-tests/hibernate-reactive-panache/providers.txt:
```diff
@@ -0,0 +1,4 @@
+io.quarkus.resteasy.runtime.ForbiddenExceptionMapper
```
Ouch, sorry.

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -0,0 +1,506 @@
+    }
+
+    public Uni<Void> persist(Object firstEntity, Object... entities) {
+        List<Object> array = new ArrayList<>(entities.length + 1);
```
I've changed this anyway.

FroMage commented on integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/TestEndpoint.java:
```diff
@@ -0,0 +1,1772 @@
+    @Path("model")
+    @Transactional
+    public Uni<String> testModel() {
+        return Panache.withTransaction(() -> Person.findAll().list()
```
Yes, transaction is required for write operations. The code is indeed hard to read, but it helped writing it to spot issues in the API, so I'd rather keep it reactive.

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -0,0 +1,506 @@
+        return count(entityClass, query, params.map());
+    }
+
+    public Uni<Boolean> exists(Class<?> entityClass) {
+        return count(entityClass).map(c -> c > 0);
+    }
+
+    public Uni<Boolean> exists(Class<?> entityClass, String query, Object... params) {
+        return count(entityClass, query, params).map(c -> c > 0);
+    }
+
+    public Uni<Boolean> exists(Class<?> entityClass, String query, Map<String, Object> params) {
+        return count(entityClass, query, params).map(c -> c > 0);
+    }
+
+    public Uni<Boolean> exists(Class<?> entityClass, String query, Parameters params) {
+        return count(entityClass, query, params).map(c -> c > 0);
+    }
```
Hum, how did we miss this, they're even missing in the ORM version…

FroMage replied:
OK, so this is used by c4e84c556b84dbbb604ea4b8bcce3b421ffc3af8 by Spring data even though we don't expose it. Mystery solved. I could remove this, but I bet when sprint data reactive lands they will need this, so…

loicmathieu replied:
Ah, OK, maybe add a small comment now that we know why it exist ;)

FroMage committed:
PanacheMock: made synchronized

FroMage commented:
Thanks everyone for the comments, I've rebased on master, and taken all comments into account. I've added the new `quarkus-panache-hibernate-common` modules to share code between orm/reactive. ATM the deployment module doesn't contain anything but I'm sure it will so I create it in advance to save refactoring later on.
Hopefully CI passes.

FroMage commented:
> Hopefully CI passes.

Famous last words. Turns out `-Dquickly` doesn't reformat :(

FroMage commented:
Fixed a copy/paste bug, let's wait for CI.

FroMage commented:
Now it's the mockito thing that doesn't pass anymore.

FroMage committed:
Fixed AlphaObserver test by making the mocked method public

Mockito doesn't know how to mock package-private methods. This only
works if we're mocking the ClientProxy class instead of the real bean
class.

FroMage commented:
New test failure appeared just after I rebased on master :(

FroMage commented:
OK, public fields are now turned into protected fields in master, fine, fixed my test.

FroMage commented:
It looks like the JDK14 test fails, but I hear it's a CI issue.

FroMage commented:
So yeah, CI passes, we can ignore JDK14.

Can we get new reviews?

FroMage commented:
@loicmathieu can I get a new review please so I can merge this?

loicmathieu commented:
@FroMage can you just explain why you need both `@Transactional` and `Panache.transaction()` ?
I didn't see anything special to handle `@Transactional` inside HIbernate Reactive with Panache ...

FroMage commented:
Rebased again, fixed compilation issues and enforcer issue.

FroMage commented:
> @FroMage can you just explain why you need both @Transactional and Panache.transaction() ?
> I didn't see anything special to handle @Transactional inside HIbernate Reactive with Panache ...

Sorry you're right that's an oversight I've removed the `@Transactional` annotations.

loicmathieu commented:
> Sorry you're right that's an oversight I've removed the @Transactional annotations.

Thanks it's clearer now.

loicmathieu reviewed (approved):
LGTM

gavinking commented:
@FroMage did you see that @Sanne just did a release of HR and it might break stuff?

FroMage commented:
> @FroMage did you see that @Sanne just did a release of HR and it might break stuff?

No, what sort of changes?

FroMage commented:
Damnit, the junit-Mockito fix of using the top class for mocks causes the IT tests for mongodb-panache-kotlin to fail with:

```
Underlying exception : java.lang.IllegalStateException: Error invoking java.lang.reflect.Executable#getParameters
	at io.quarkus.test.junit.mockito.internal.CreateMockitoMocksCallback.createMockAndSetTestField(CreateMockitoMocksCallback.java:44)
	at io.quarkus.test.junit.mockito.internal.CreateMockitoMocksCallback.beforeAll(CreateMockitoMocksCallback.java:28)
	... 67 more
Caused by: java.lang.IllegalStateException: Error invoking java.lang.reflect.Executable#getParameters
	at net.bytebuddy.description.method.ParameterDescription$ForLoadedParameter$Dispatcher$ForJava8CapableVm.getParameter(ParameterDescription.java:523)
	at net.bytebuddy.description.method.ParameterDescription$ForLoadedParameter$Dispatcher$ForJava8CapableVm.isNamePresent(ParameterDescription.java:489)
	at net.bytebuddy.description.method.ParameterDescription$ForLoadedParameter.isNamed(ParameterDescription.java:277)
	at net.bytebuddy.description.method.ParameterDescription$TypeSubstituting.isNamed(ParameterDescription.java:1080)
	at net.bytebuddy.description.method.ParameterDescription$AbstractBase.asToken(ParameterDescription.java:181)
	at net.bytebuddy.description.method.ParameterDescription$AbstractBase.asToken(ParameterDescription.java:129)
	at net.bytebuddy.description.method.ParameterList$AbstractBase.asTokenList(ParameterList.java:98)
	at net.bytebuddy.description.method.MethodDescription$AbstractBase.asToken(MethodDescription.java:801)
	at net.bytebuddy.description.method.MethodDescription$AbstractBase.asToken(MethodDescription.java:373)
	at net.bytebuddy.dynamic.Transformer$ForMethod.transform(Transformer.java:298)
	at net.bytebuddy.dynamic.Transformer$ForMethod.transform(Transformer.java:253)
	at net.bytebuddy.dynamic.Transformer$Compound.transform(Transformer.java:645)
	at net.bytebuddy.dynamic.scaffold.MethodRegistry$Default$Entry.asPreparedEntry(MethodRegistry.java:596)
	at net.bytebuddy.dynamic.scaffold.MethodRegistry$Default.prepare(MethodRegistry.java:484)
	at net.bytebuddy.dynamic.scaffold.subclass.SubclassDynamicTypeBuilder.make(SubclassDynamicTypeBuilder.java:213)
	at net.bytebuddy.dynamic.scaffold.subclass.SubclassDynamicTypeBuilder.make(SubclassDynamicTypeBuilder.java:204)
	at net.bytebuddy.dynamic.DynamicType$Builder$AbstractBase.make(DynamicType.java:3595)
	at org.mockito.internal.creation.bytebuddy.SubclassBytecodeGenerator.mockClass(SubclassBytecodeGenerator.java:199)
	at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator$1.call(TypeCachingBytecodeGenerator.java:46)
	at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator$1.call(TypeCachingBytecodeGenerator.java:43)
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:152)
	at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:365)
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:174)
	at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:376)
	at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator.mockClass(TypeCachingBytecodeGenerator.java:36)
	at org.mockito.internal.creation.bytebuddy.SubclassByteBuddyMockMaker.createMockType(SubclassByteBuddyMockMaker.java:77)
	at org.mockito.internal.creation.bytebuddy.SubclassByteBuddyMockMaker.createMock(SubclassByteBuddyMockMaker.java:43)
	at org.mockito.internal.creation.bytebuddy.ByteBuddyMockMaker.createMock(ByteBuddyMockMaker.java:29)
	at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:52)
	at org.mockito.internal.MockitoCore.mock(MockitoCore.java:61)
	at org.mockito.Mockito.mock(Mockito.java:1949)
	at org.mockito.Mockito.mock(Mockito.java:1860)
	... 69 more
Caused by: java.lang.reflect.MalformedParametersException: Invalid parameter name ""
	at java.lang.reflect.Executable.verifyParameters(Executable.java:386)
	at java.lang.reflect.Executable.privateGetParameters(Executable.java:416)
	at java.lang.reflect.Executable.getParameters(Executable.java:357)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.bytebuddy.description.method.ParameterDescription$ForLoadedParameter$Dispatcher$ForJava8CapableVm.getParameter(ParameterDescription.java:519)
	... 100 more

```

And

```
[ERROR] io.quarkus.it.mongodb.panache.MongodbPanacheMockingTest.testPanacheRepositoryBridges  Time elapsed: 0.007 s  <<< ERROR!
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
	at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:638)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:615)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.api.extension.InvocationInterceptor.interceptTestClassConstructor(InvocationInterceptor.java:72)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:77)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeTestClassConstructor(ClassBasedTestDescriptor.java:333)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateTestClass(ClassBasedTestDescriptor.java:280)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:77)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateAndPostProcessTestInstance(ClassBasedTestDescriptor.java:262)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$2(ClassBasedTestDescriptor.java:256)
	at java.util.Optional.orElseGet(Optional.java:267)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$3(ClassBasedTestDescriptor.java:255)
	at org.junit.jupiter.engine.execution.TestInstancesProvider.getTestInstances(TestInstancesProvider.java:29)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$prepare$0(TestMethodTestDescriptor.java:108)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:107)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:71)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$1(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:75)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:248)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$5(DefaultLauncher.java:211)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:226)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:199)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:132)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:188)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:128)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:635)
	... 62 more
Caused by: org.mockito.exceptions.base.MockitoException: 
Mockito cannot mock this class: class io.quarkus.it.mongodb.panache.person.MockablePersonRepository.

Mockito can only mock non-private & non-final classes.
If you're not sure why you're getting this error, please report to the mailing list.

```

So essentially, ArC is making Kotlin classes mockable by accident, because the generated Proxy class is valid (even though it has no generics info) where the original Kotlin classes are not, or at least trigger bugs in bytebuddy.

@evanchooly or @mkouba do you know how I can proceed from there?

FroMage commented:
@mkouba OK I've had it with this delaying my PR, I'm implementing signatures for the proxies. Resolves my issue, now let's see if it breaks anything else.

mkouba commented:
@FroMage I'm -0 or even -1 for implementing signatures for the proxies. It could cause various issues here and there and I doubt our test suite is robust enough to catch them all. And the only use case needed seems to be the panache kotlin reactive hibernate mockito combo. Is that worth the effort? BTW the same problem should arise with intercepted subclasses, i.e. constructs generated for intercepted beans.

FroMage committed:
Added AsmUtilCopy to ArC and use it to create signatures in _ClientProxy classes

FroMage commented:
Well, I agree I would rather not need this, but the fact is that:

- ClientProxy classes are more public by default and have a default constructor, which helps Mockito
- I can't figure out how to fix the Kotlin stuff, but I do know how to generate signatures
- Ultimately, the JVM spec does say that if your method/class have generics, they should have signatures, so this is technically more correct, even if this is the first time we've needed this

Now, if @evanchooly can figure out how to fix the Kotlin byte code, I'm ready to change side until a better reason comes up to add signatures, but I really hate how this is blocking merging this PR for reasons that are totally unrelated to the core PR.

I've added the fix to the PR for now, let's at least see if CI passes.

FroMage commented:
@mkouba if CI passes, perhaps you can check out https://github.com/quarkusio/quarkus/pull/10769/commits/f3af0b36effdfc439ecba7b8fca91ed50c8fa116 and see if it's obviously horrlble or not?

FroMage commented:
FTR, the JVM spec says this about signatures (see https://docs.oracle.com/javase/specs/jvms/se14/html/jvms-4.html#jvms-4.7.9.1):

> Signatures encode declarations written in the Java programming language that use types outside the type system of the Java Virtual Machine. They support reflection and debugging, as well as compilation when only class files are available.

> A Java compiler must emit a signature for any class, interface, constructor, method, or field whose declaration uses type variables or parameterized types

We can argue whether this applies to ArC, which is not a compiler, but it's supposed to produce valid classes.

And:

> Oracle's Java Virtual Machine implementation does not check the well-formedness of Signature attributes during class loading or linking. Instead, Signature attributes are checked by methods of the Java SE Platform class libraries which expose generic signatures of classes, interfaces, constructors, methods, and fields. Examples include getGenericSuperclass in Class and toGenericString in java.lang.reflect.Executable.

Which says that this shows up only when doing reflection on objects, which is the case here.

evanchooly commented:
What's missing from the generated kotlin bits?  I can take a look but the top level class signatures are just passed in from what's read off disk.  I don't think anything touches those.

FroMage commented:
It's not the signatures, it's the method parameter attributes that are wrong, if you check the first exception. The second exception I've no idea what it's about.

FroMage commented:
I tried debugging this, but it doesn't look like `quarkus.debug.transformed-classes-dir` is enough to make the transformed classes appear.
It appears to be the method proxies that have empty parameter names, stuff like `MockagePersonRepository.count(String,Map)` which we copy from the supertype.
But I can't look at the actual bytes.

Well, I do see the class in `target/classes` but I'm not sure if it's the transformed one or not. It does look like it is because it has many methods, but with Kotlin, who knows:

```
  public long count(java.lang.String, java.util.Map<java.lang.String, ? extends java.lang.Object>);
    descriptor: (Ljava/lang/String;Ljava/util/Map;)J
    flags: ACC_PUBLIC
    Code:
      stack=3, locals=3, args_size=3
         0: aload_1
         1: ldc           #63                 // String query
         3: invokestatic  #66                 // Method kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull:(Ljava/lang/Object;Ljava/lang/String;)V
         6: aload_2
         7: ldc           #68                 // String params
         9: invokestatic  #66                 // Method kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull:(Ljava/lang/Object;Ljava/lang/String;)V
        12: aload_0
        13: aload_1
        14: aload_2
        15: invokestatic  #83                 // Method io/quarkus/mongodb/panache/kotlin/PanacheMongoRepositoryBase$DefaultImpls.count:(Lio/quarkus/mongodb/panache/kotlin/PanacheMongoRepositoryBase;Ljava/lang/String;Ljava/util/Map;)J
        18: lreturn
      LineNumberTable:
        line 8: 12
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      19     0  this   Lio/quarkus/mongodb/panache/kotlin/PanacheMongoRepositoryBase;
            0      19     1 query   Ljava/lang/String;
            0      19     2 params   Ljava/util/Map;
    Signature: #80                          // (Ljava/lang/String;Ljava/util/Map<Ljava/lang/String;+Ljava/lang/Object;>;)J
    RuntimeVisibleAnnotations:
      0: #53()
    RuntimeInvisibleParameterAnnotations:
      0:
        0: #61()
      1:
        0: #61()
```

Which looks OK.

FroMage commented:
@mkouba do you know if we can actually dump the transformed classes? I can't find any reference to `DEBUG_TRANSFORMED_CLASSES_DIR` that would use it.

evanchooly commented:
i wrote a script to `javap` the classes dir and then `javap` the `-runner` jar so i can see the before/after on the generated bytecode, fwiw.  It does a bit more than that but I can share it if you'd like.  But it's pretty simple to update the first `javap` to use the jar instead.  my script just adds some rebuilding steps first to make sure it's running the latest code.

mkouba commented:
> @mkouba do you know if we can actually dump the transformed classes? I can't find any reference to DEBUG_TRANSFORMED_CLASSES_DIR that would use it.

It seems that Stuart removed it in this commit: https://github.com/quarkusio/quarkus/commit/9d3fa16a74c1b9354352843aab28ce71b3c95350#diff-db1d39a812769ff88e14c8dd8e39c0c4 :-(

@stuartwdouglas Any particular reason why?

FroMage commented:
So, in IT, what will give me a jar? `mvn package` won't.

FroMage commented:
@stuartwdouglas is on PTO this week. I'm just going to sit here and cry.

evanchooly commented:
> So, in IT, what will give me a jar? `mvn package` won't.

`mvn package -Dquarkus.build.skip=false -DskipTests=true`

FroMage commented:
OK thanks.

This validates that the output is broken:

```
  public long count(java.lang.String, java.lang.Object[]);
    descriptor: (Ljava/lang/String;[Ljava/lang/Object;)J
    flags: ACC_PUBLIC
    Code:
      stack=4, locals=3, args_size=3
         0: getstatic     #403                // Field io/quarkus/mongodb/panache/kotlin/PanacheMongoEntityBase.Companion:Lio/quarkus/mongodb/panache/kotlin/PanacheMongoEntityBase$Companion;
         3: invokevirtual #409                // Method io/quarkus/mongodb/panache/kotlin/PanacheMongoEntityBase$Companion.getOperations:()Lio/quarkus/mongodb/panache/kotlin/runtime/KotlinMongoOperations;
         6: ldc           #94                 // class io/quarkus/it/mongodb/panache/person/PersonEntity
         8: aload_1
         9: aload_2
        10: invokevirtual #417                // Method io/quarkus/mongodb/panache/kotlin/runtime/KotlinMongoOperations.count:(Ljava/lang/Class;Ljava/lang/String;[Ljava/lang/Object;)J
        13: lreturn
    Signature: #74                          // (Ljava/lang/String;[Ljava/lang/Object;)J
    MethodParameters:
      Name                           Flags
      <no name>
      <no name>

```

FroMage commented:
So, something in the kotlin transformer adds method parameter names where there were none before, but with broken names.

FroMage commented:
FTR, the same `integration-tests/mongodb-panache` test has this output:

```
  public long count(java.lang.String, java.lang.Object[]);
    descriptor: (Ljava/lang/String;[Ljava/lang/Object;)J
    flags: ACC_PUBLIC
    Code:
      stack=4, locals=3, args_size=3
         0: getstatic     #51                 // Field operations:Lio/quarkus/mongodb/panache/runtime/MongoOperations;
         3: ldc           #53                 // class io/quarkus/it/mongodb/panache/person/PersonEntity
         5: aload         1
         7: aload         2
         9: invokevirtual #64                 // Method io/quarkus/mongodb/panache/runtime/MongoOperations.count:(Ljava/lang/Class;Ljava/lang/String;[Ljava/lang/Object;)J
        12: lreturn
    Signature: #59                          // (Ljava/lang/String;[Ljava/lang/Object;)J
    MethodParameters:
      Name                           Flags
      query
      params
```

So here the names are not all wrong.

FroMage commented:
Perhaps because the source parameters are present for the Java IT but not for the Kotlin IT in the source method?

Yes, `io.quarkus.mongodb.panache.PanacheMongoRepositoryBase` has parameter names but `io.quarkus.mongodb.panache.kotlin.PanacheMongoRepositoryBase` does not.

FroMage commented:
OK, compiling with parameter names fixes this test, apparently, which means I can go back to removing signatures from ArC.

But I'm still not sure I should…

evanchooly commented:
what changes did you have to make so I can make them in my current code as well?  was it a code change or a build change?

FroMage committed:
Make the Kotlin MongoDB with Panache tests pass

FroMage commented:
I've pushed it here. But I will add a safer commit tomorrow that is less fragile for all versions of the panache enhancers.

evanchooly commented:
simple enough.  shouldn't conflict with what I have in motion.  fwiw, I'm working on reducing the number of those enhancers/visitors which will reduce the maintenance burden there.

FroMage committed:
Factored all visitParameter into AsmUtil to better avoid missing parameter names

FroMage commented:
Added a new method to guard against future copying of null parameter names, and rebased to resolve HR conflicts.

FroMage commented:
Fixed the damn formatting.

FroMage commented:
All the failures are unrelated:

```
2020-09-01T11:48:45.8896837Z Resulted in: org.jboss.resteasy.spi.UnhandledException: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: http://localhost:40435/api/v1/namespaces/test/pods.
```

Let's restart.

stuartwdouglas commented:
Removing the dump must have been an oversight.

FroMage commented:
Rebased to get rid of kubernetes test failing.

jaikiran commented:
Failure in the `IncompletePostTestCase` is being tracked at https://github.com/quarkusio/quarkus/issues/11839

FroMage commented:
Let's rebase.

FroMage commented:
Yay!

Sanne commented:
Wow! very nice

loicmathieu commented:
Cool ! I need to test this !

gavinking commented:
Great!

yuhaibohotmail commented:
@FroMage ,this PR has include in 19.CR1, Could you provide some documents for this PR?
